### PR TITLE
Python console top level window

### PR DESCRIPTION
**Pull request by stevenmizuno**
This PR provides a way to use Python Console as either a dock widget or a top-level window. It can be switched on the fly - no need to restart QGIS.

see Feature Request https://issues.qgis.org/issues/4019

Please assign this PR to Borys Jurgiel. 

The basics to switch from one type to the other:
on change of type - determined by the current state not being the same as the state stored previously
1. construct a new widget of the other type
2. re-parent the console widget by setting it to the new widget
3. remove/delete the existing window object that contained the console
4. if the new widget is a dockwidget add it to the main window
5. save the new state

Changing Python Console's window type takes effect the next time Plugins, Python Console is chosen, even if the console is visible. As the editor is retained, nothing done in the console will be lost.

I have chosen to use the setting qgis/dockPythonConsole with true for dockwidget (also the default without the setting being present), false for top-level window.

When it is a dock window it is alternately shown/hidden each time the Plugins, Python Console menu item is chosen; and it will appear in the main window's context menu and can be toggled on and off there (no change from present behavior).

As a top-level window it will be shown each time the Plugins, Python Console menu item is chosen; the window may be minimized or closed using the buttons on its title bar. And you can navigate to the console window using Alt+Tab.
It is possible to make the window be hidden/shown on alternate operation of the Python Console menu item, but I felt it better to always show the console as it may be behind other windows or minimized. If hide() is called on the console window the window won't be listed as a window of the application on the task bar and you can't navigate to it using Alt+Tab. This may differ depending on the operating system.

The top-level window geometry is saved when the window is closed and restored when the window is constructed.

I'm not sure whether or where to put an option toggle (checkbox) in QGIS Settings|Options as there are no longer such settings for windows.

I have considered putting the setting for top-level vs. dockwidget somewhere in Python Console itself. This would make this feature totally contained in the console code, rather than have some part in the main QGIS application.

Please provide some guidance on this. 

Some sample Python code (use in the console):

<pre># create the setting for dockwidget or top-level window
s=QgsSettings()

s.setValue('qgis/dockPythonConsole','true')	# dockwidget
# or
s.setValue('qgis/dockPythonConsole','false')	# top-level

# demonstrate hiding window, especially as a top-level window
import console
c=console.show_console()

c.hide()
</pre>

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: python/console/console.py [hunk 1]
```diff
@@ -37,22 +37,64 @@
...
-    if _console is None:
-        parent = iface.mainWindow() if iface else None
-        _console = PythonConsole(parent)
+    global _consoleEditor
+    global _consoleAsDock
+
+    s = QgsSettings()
+
+    # get the setting as a bool value (for "true"/"false" values QgsSettings returns a string to Python)
+    dockValue = QVariant(s.value("qgis/dockPythonConsole","true"))
```
Comment (borysiasty) [inline python/console/console.py]: You didn't run scripts/prepare-commit.sh ;) There is a missing space after the comma - although I don't think the failing test is related to your commit.

File: python/console/console.py [hunk 2]
```diff
@@ -96,6 +140,33 @@ def closeEvent(self, event):
         QWidget.closeEvent(self, event)
 
 
+class PythonConsoleW(QWidget):
+    """ as a QWidget """
```
Comment (borysiasty) [inline python/console/console.py]: Astyle also complains about missing empty line between the comment and the method definition.

**Comment by borysiasty** [comment]
> As a top-level window it will be shown each time the Plugins, Python Console menu item is chosen; the window may be minimized or closed using the buttons on its title bar. And you can navigate to the console window using Alt+Tab.
> It is possible to make the window be hidden/shown on alternate operation of the Python Console menu item, but I felt it better to always show the console as it may be behind other windows or minimized. If hide() is called on the console window the window won't be listed as a window of the application on the task bar and you can't navigate to it using Alt+Tab. This may differ depending on the operating system.

Personally, I don't like the fact the menu action acts differently in both modes. Despite your arguments above, I'd use the show/hide approach also in the window mode - for sake of consistency. If the window is minimised or hidden behind, I can click the action twice and blame myself I didn't notice it earlier.

**Comment by borysiasty** [comment]
> As a top-level window it will be shown each time the Plugins, Python Console menu item is chosen; the window may be minimized or closed using the buttons on its title bar. And you can navigate to the console window using Alt+Tab.
> It is possible to make the window be hidden/shown on alternate operation of the Python Console menu item, but I felt it better to always show the console as it may be behind other windows or minimized. If hide() is called on the console window the window won't be listed as a window of the application on the task bar and you can't navigate to it using Alt+Tab. This may differ depending on the operating system.

Personally, I don't like the fact the menu action acts differently in both modes. Despite your arguments above, I'd use the show/hide approach also in the window mode - for sake of consistency. If the window is minimised or hidden behind, I can click the action twice and blame myself I didn't notice it earlier.

**Comment by borysiasty** [comment]
Wow, the patch is back. It disappeared magically (together with your user account) the week ago when I was trying to post this comment :) Seems github has some issues.

> I'm not sure whether or where to put an option toggle (checkbox) in QGIS Settings|Options as there 
> are no longer such settings for windows.

The setting for the attribute table was just moved to more dedicated place: Options|Data Sources. So yes, there is no general place for such switches any longer.

> I have considered putting the setting for top-level vs. dockwidget somewhere in Python Console itself.
> This would make this feature totally contained in the console code, rather than have some part in the 
> main QGIS application.

The only place I see is the Console's settings dialog. It doesn't fit to any of the existing tabs, so a new "General" tab with this single switch would be necessary. Maybe not for long - there was a discussion about possible moving the console options to the main options dialog, so one day the tabs may be downgraded to groups again: https://github.com/qgis/QGIS/pull/4550

**Comment by stevenmizuno** [comment]
On your comment about the menu action behaving differently on the widget types, I don't have strong feelings about having the top-level window always be shown. So I am OK with changing to show/hide for both ways and will submit a further commit implementing that.

For the option toggle I will try adding a General tab with a checkbox  to the console's settings.

**Comment by borysiasty** [comment]
Thanks a lot. I'm setting it for QGIS 3.2, as we're in the hard feature freeze now - I hope I'm not making you sad you're not disappointed it has to wait a month to be pulled! :)

**Comment by stevenmizuno** [comment]
No problem - I expected that it might not be put it just now.

**Comment by m-kuhn** [comment]
Would it be a good time to revive this? @stevenmizuno @borysiasty

**Comment by borysiasty** [comment]
Sure! @stevenmizuno  I don't want to hurry you, but when you have some time, it would be great to complete it.

File: python/console/console.py [hunk 3]
```diff
@@ -96,6 +135,33 @@ def closeEvent(self, event):
         QWidget.closeEvent(self, event)
 
 
+class PythonConsoleW(QWidget):
```
Comment (m-kuhn) [inline python/console/console.py]: Better write this completely (`PythonConsoleWidget` or whatever it's meant to be, so we do not mistake it with `PythonConsoleWinterfell`, `PythonConsoleWhistleblower` or `PythonConsoleWindow`)

File: python/console/console.py [hunk 4]
```diff
@@ -37,22 +37,59 @@
...
 
 
 def show_console():
     """ called from QGIS to open the console """
     global _console
-    if _console is None:
-        parent = iface.mainWindow() if iface else None
-        _console = PythonConsole(parent)
+    global _consoleEditor
+    global _consoleAsDock
```
Comment (m-kuhn) [inline python/console/console.py]: Is there any chance to use something else than globals here?

File: python/console/console.py [hunk 5]
```diff
@@ -37,22 +37,59 @@
...
+
+    s = QgsSettings()
+
+    # get the setting as a bool value (for "true"/"false" values QgsSettings returns a string to Python)
+    dockValue = QVariant(s.value("qgis/dockPythonConsole", "true"))
```
Comment (m-kuhn) [inline python/console/console.py]: I think that should work as well

```
consoleAsDock = s.value(("qgis/dockPythonConsole", true)
```

**Comment by m-kuhn** [comment]
Good job.
Something that could be interesting for the window mode here is the [`Qt.WindowType` `Qt.Tool`](http://doc.qt.io/qt-5/qt.html#WindowType-enum). This will make it float always on top of the application without docking. Not sure if it's better to make it always on top or not, your call.

File: python/console/console.py (hunk 5 above)
Comment (m-kuhn) [reply python/console/console.py]: where is this set?

**Comment by nyalldawson** [comment]
Ping @probot

**Pull request closed without merging**